zulip_bots/zulip_bots/bots/monkeytestit/libraries/report.py

```diff
@@ -0,0 +1,130 @@
+"""Used to mainly compose a decorated report for the user
+"""
+
+from typing import Dict, Text, List
+
+
+def compose(results: Dict) -> Text:
+    """Composes a report based on test results
+
+    An example would be:
+
+    Status: tests_failed
+    Enabled checkers: seo
+    Failures from checkers: seo (3)
+    More info: https://monkeytest.it/...
+
+    This function assumes that the result dict is valid and does not contain
+    any "errors" like bad url
+
+    :param results: A dictionary containing the results of a check
+    :return: A response string containing the full report
+    """
+    if "error" in results:
+        return "Error: {}".format(results['error'])
+
+    response = ""
+
+    response += "{}\n".format(print_status(results))
+
+    if "success" in response.lower():
+        response += "{}".format(print_test_id(results))
+        return response
+
+    response += "{}\n".format(print_enabled_checkers(results))
+    response += "{}\n".format(print_failures_checkers(results))
+    response += "{}".format(print_more_info_url(results))
+
+    return response
+
+
+def print_more_info_url(results: Dict) -> Text:
+    """Creates info for the test URL from monkeytest.it
+
+    Example:
+
+    More info: https://monkeytest.it/test/...
+
+    :param results: A dictionary containing the results of a check
+    :return: A response string containing the url info
+    """
+    return "More info: {}".format(results['results_url'])
+
+
+def print_test_id(results: Dict) -> Text:
+    """Prints the test-id with attached to the url
+
+    :param results: A dictionary containing the results of a check
+    :return: A response string containing the test id
+    """
+    return "Test: https://monkeytest.it/test/{}".format(results['test_id'])
+
+
+def print_failures_checkers(results: Dict) -> Text:
+    """Creates info for failures in enabled checkers
+
+    Example:
+
+    Failures from checkers: broken_links (3), seo (5)
+
+    This means that the check has 7 section failures, 3 sections in
+    broken_links and the other 5 are in seo.
+
+    :param results: A dictionary containing the results of a check
+    :return: A response string containing number of failures in each enabled
+             checkers
+    """
+    failures_checkers = [(checker, len(results['failures'][checker]))
+                         for checker in get_enabled_checkers(results)
+                         if checker in results['failures']]  # [('seo', 3), ..]
+
+    failures_checkers_message = ""
+
+    for fail_checker in failures_checkers:
+        failures_checkers_message += "{} ({}), ".format(fail_checker[0],
+                                                        fail_checker[1])
+
+    failures_checkers_message = failures_checkers_message[:-2]  # "a, " -> "a"
```

Comment:
You can rephrase lines 83-87 with `join` statements and list comprehensions:

fail_messages = [' '.join(items[:2] for items in failures_checkers]
big_message = ', '.join(fail_messages)